### PR TITLE
Base block GUI

### PR DESCRIPTION
**Pull request by juanmuscaria**
Currently, a WIP of the base block GUI.
Living and entity handlers were modified to add the ability to listen to block place and break.
ViewScreen got an extra constructor to define whenever the screen should pause the came, and using the original constructor will pause the game by default as it was the original behavior before the change

### CODE REVIEW

File: crafting-dead-core/src/main/java/com/craftingdead/core/CraftingDead.java [hunk 1]
```diff
@@ -307,6 +311,34 @@ public void handleLivingDamage(LivingDamageEvent event) {
...
+      event.getEntity()
+          .getCapability(LivingExtension.CAPABILITY)
+          .ifPresent(living -> event.setCanceled(
+              living.handleMultiBlockPlace(event.getReplacedBlockSnapshots(), event.getPlacedBlock(), event.getPlacedAgainst())));
+    }
+  }
+
+  @SubscribeEvent(priority = EventPriority.LOWEST)
+  public void handleEntityBlockBreakEvent(BlockEvent.BreakEvent event) {
+    // java lacks the ability to pass primitives by reference, with mutable integer whe can emulate that behavior
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/CraftingDead.java]: Instead, add an `IntConsumer` parameter and pass it `event::setExpToDrop`

File: crafting-dead-core/src/main/java/com/craftingdead/core/world/item/ModItems.java [hunk 2]
```diff
@@ -2324,6 +2326,11 @@ public ItemStack makeIcon() {
               .stacksTo(1)
               .tab(COMBAT_TAB)));
 
+  public static final RegistryObject<Item> MIMIC_BLOCK_ITEM =
+      ITEMS.register("mimi",
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/world/item/ModItems.java]: Typo

File: crafting-dead-immerse/src/main/java/com/craftingdead/immerse/client/gui/screen/block/base/ManageBaseMembersView.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+  private final ParentView memberList;
+  private final TextFieldView newMemberField;
+  private final LegacyBase base;
+  private final Map<UUID, String> usernames;
+  /**
+  0 - remove button
+  1 - promote button
+  2 - demote button
+   */
+  private final View[] memberListButtons = new View[3];
```
Comment (josephtarbit) [inline crafting-dead-immerse/src/main/java/com/craftingdead/immerse/client/gui/screen/block/base/ManageBaseMembersView.java]: Use separate fields instead of an array

File: crafting-dead-immerse/src/main/java/com/craftingdead/immerse/game/survival/SurvivalPlayerHandler.java [hunk 4]
```diff
@@ -93,6 +99,17 @@ public boolean handleKill(Entity target) {
     return false;
   }
 
+  @Override
+  public boolean handleBlockBreak(BlockPos pos, BlockState block, MutableInt xp) {
+    if (block.is(ImmerseBlocks.BASE_CENTER.get())
+        && LevelExtension.getOrThrow(this.player.getLevel()).getLandManager().getLandOwnerAt(pos).orElse(null) instanceof LegacyBase base
+        && !base.getOwnerId().equals(this.player.getEntity().getUUID())) {
```
Comment (josephtarbit) [inline crafting-dead-immerse/src/main/java/com/craftingdead/immerse/game/survival/SurvivalPlayerHandler.java]: Check that they're within the build range of the base too

File: crafting-dead-immerse/src/main/java/com/craftingdead/immerse/network/play/BaseScreenActionMessage.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    var missingId = this.targetProfile.getId() == null;
+    out.writeBoolean(missingId);
+    if (!missingId) {
+      out.writeUUID(this.targetProfile.getId());
+    }
+  }
+
+  public boolean handle(Supplier<Context> ctx) {
```
Comment (josephtarbit) [inline crafting-dead-immerse/src/main/java/com/craftingdead/immerse/network/play/BaseScreenActionMessage.java]: Don't import nested classes, use `NetworkEvent.Context`

File: crafting-dead-immerse/src/main/java/com/craftingdead/immerse/network/play/SyncBaseScreenMessage.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            .map(LevelExtension::getLandManager)
+            .ifPresent(landManager -> {
+              var land = landManager.getLandOwner(this.base);
+              if (land instanceof LegacyBase legacyBase) {
+                Minecraft.getInstance().setScreen(
+                    ManageBaseMembersView.createScreen(legacyBase, this.usernames));
+              }
+            });
+        case SYNC_MEMBERS, SYNC_AND_CLEAR -> {
+          if (Minecraft.getInstance().screen instanceof ViewScreen view
```
Comment (josephtarbit) [inline crafting-dead-immerse/src/main/java/com/craftingdead/immerse/network/play/SyncBaseScreenMessage.java]: This may crash dedicated servers, if it does, move it into a separate static method which will stop `Minecraft` being class loaded

**Review by josephtarbit — CHANGES_REQUESTED** [review]

File: crafting-dead-immerse/src/main/java/com/craftingdead/immerse/network/play/SyncBaseScreenMessage.java (hunk 6 above)
Comment (juanmuscaria) [reply crafting-dead-immerse/src/main/java/com/craftingdead/immerse/network/play/SyncBaseScreenMessage.java]: I assume when registering a packet, using `NetworkDirection.PLAY_TO_CLIENT` indicates that the packet will only ever be processed by the client, so it should be safe to do any client action there right?

File: crafting-dead-immerse/src/main/java/com/craftingdead/immerse/game/survival/SurvivalPlayerHandler.java (hunk 4 above)
Comment (juanmuscaria) [reply crafting-dead-immerse/src/main/java/com/craftingdead/immerse/game/survival/SurvivalPlayerHandler.java]: Is it needed in the block break event? Minecraft already checks if the block is reachable before allowing it to be broken if I'm not mistaken.

File: crafting-dead-core/src/main/java/com/craftingdead/core/CraftingDead.java (hunk 1 above)
Comment (juanmuscaria) [reply crafting-dead-core/src/main/java/com/craftingdead/core/CraftingDead.java]: Having only an IntConsumer will remove the ability of the handles to be aware of the original value (usually applied by fortune and some block that drops xp, some other event or handler that modified the xp), and having a Supplied and Consumer pair makes it awkward to work with getting and setting the xp.

File: crafting-dead-immerse/src/main/java/com/craftingdead/immerse/game/survival/SurvivalPlayerHandler.java (hunk 4 above)
Comment (josephtarbit) [reply crafting-dead-immerse/src/main/java/com/craftingdead/immerse/game/survival/SurvivalPlayerHandler.java]: Yeah because there’s only a certain area you can modify blocks with the base claim. The inner area is buildable but the outer area is not (acts as a buffer zone between bases).

File: crafting-dead-immerse/src/main/java/com/craftingdead/immerse/network/play/SyncBaseScreenMessage.java (hunk 6 above)
Comment (josephtarbit) [reply crafting-dead-immerse/src/main/java/com/craftingdead/immerse/network/play/SyncBaseScreenMessage.java]: Not necessarily as the message class will still be class loaded on the server which could cause a crash.

File: crafting-dead-core/src/main/java/com/craftingdead/core/CraftingDead.java (hunk 1 above)
Comment (josephtarbit) [reply crafting-dead-core/src/main/java/com/craftingdead/core/CraftingDead.java]: That’s true, never mind then.

File: crafting-dead-immerse/src/main/java/com/craftingdead/immerse/game/survival/SurvivalPlayerHandler.java (hunk 4 above)
Comment (juanmuscaria) [reply crafting-dead-immerse/src/main/java/com/craftingdead/immerse/game/survival/SurvivalPlayerHandler.java]: The current code also has an unintended side effect of all base blocks without a land be only breakable by the owner of the valid base, I changed the way I grab the base owner to check but it looks a little over complicated

**Comment by josephtarbit** [comment]
Merge upstream changes and then I'll review.

**Comment by 0xE69** [comment]
There seems to be a conflicting file here; crafting-dead-immerse/src/main/java/sm0keysa1m0n/bliss/platform/ScissorStack.java

File: crafting-dead-immerse/src/main/java/com/craftingdead/immerse/client/gui/screen/block/base/ManageBaseMembersView.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+  private final ParentView addMemberView;
+  private final Map<UUID, String> usernames;
+  private String lastSearch = "";
+
+  public ManageBaseMembersView(LegacyBase base, Map<UUID, String> usernames) {
+    super(new Properties());
+    this.base = base;
+    this.usernames = usernames;
+    this.memberList = new ParentView(new Properties().id("member-list")) {
+      // Work around list items making the search field unclickable
```
Comment (josephtarbit) [inline crafting-dead-immerse/src/main/java/com/craftingdead/immerse/client/gui/screen/block/base/ManageBaseMembersView.java]: Why is this?

File: crafting-dead-immerse/src/main/java/sm0keysa1m0n/bliss/layout/yoga/YogaLayout.java [hunk 8]
```diff
@@ -498,6 +503,12 @@ public void markDirty() {
     Yoga.YGNodeMarkDirty(this.node);
   }
 
+  @Override
+  public void clearStyles() {
```
Comment (josephtarbit) [inline crafting-dead-immerse/src/main/java/sm0keysa1m0n/bliss/layout/yoga/YogaLayout.java]: Why is this required?

**Review by josephtarbit — CHANGES_REQUESTED** [review]

File: crafting-dead-immerse/src/main/java/com/craftingdead/immerse/client/gui/screen/block/base/ManageBaseMembersView.java (hunk 7 above)
Comment (juanmuscaria) [reply crafting-dead-immerse/src/main/java/com/craftingdead/immerse/client/gui/screen/block/base/ManageBaseMembersView.java]: When you scroll, a list item will make the field unclickable due to being over it even though it's not visible, I was not sure how to make it take into account whenever the element is visible or outside the view so I made it always take priority on a click

File: crafting-dead-immerse/src/main/java/sm0keysa1m0n/bliss/layout/yoga/YogaLayout.java (hunk 8 above)
Comment (juanmuscaria) [reply crafting-dead-immerse/src/main/java/sm0keysa1m0n/bliss/layout/yoga/YogaLayout.java]: When you set a layout display to none all it's children will properly update it's size due to their layout being nil, but the parent layout still have a yoga layout, and due to styles usually assigning a fixed width and height as well other styles, the parent will still be visible, by clearing the styles for when the display is set to none will effectively make it invisible since yoga will finally update the width and height to 0.
I couldn't find a way to make yoga ignore the set width and height for when the display is none
